File: Llama3.CSharp/Sampler.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Llama3.CSharp;

public class Sampler
{
    private readonly Random _rng;

    public Sampler(int? seed = null)
    {
        _rng = seed.HasValue ? new Random(seed.Value) : new Random();
    }

    public int Sample(float[] logits, float temp, float topp)
    {
        if (temp == 0) return Argmax(logits);

        for (int i = 0; i < logits.Length; i++) logits[i] /= temp;
        MathKernels.Softmax(logits);

        if (topp <= 0 || topp >= 1) return CategoricalSample(logits);
        return ToppSample(logits, topp);
    }

    private int Argmax(float[] logits)
    {
        int maxIdx = 0;
        float maxVal = logits[0];
        for (int i = 1; i < logits.Length; i++)
        {
            if (logits[i] > maxVal)
            {
                maxVal = logits[i];
                maxIdx = i;
            }
        }
        return maxIdx;
    }

    private int CategoricalSample(float[] probs)
    {
        float r = (float)_rng.NextDouble();
        float cdf = 0;
        for (int i = 0; i < probs.Length; i++)
        {
            cdf += probs[i];
            if (r < cdf) return i;
        }
        return probs.Length - 1;
    }

    private int ToppSample(float[] probs, float topp)
    {
        float threshold = (1.0f - topp) / probs.Length;
        var candidates = new List<(float p, int i)>();
        for (int i = 0; i < probs.Length; i++)
        {
            if (probs[i] > threshold) candidates.Add((probs[i], i));
        }

        if (candidates.Count == 0) return Argmax(probs);

        candidates.Sort((a, b) => b.p.CompareTo(a.p));

        float cumulativeProb = 0;
        int lastIdx = candidates.Count - 1;
        for (int i = 0; i < candidates.Count; i++)
        {
            cumulativeProb += candidates[i].p;
            if (cumulativeProb > topp)
            {
                lastIdx = i;
                break;
            }
        }

        float r = (float)_rng.NextDouble() * cumulativeProb;
        float cdf = 0;
        for (int i = 0; i <= lastIdx; i++)
        {
            cdf += candidates[i].p;
            if (r < cdf) return candidates[i].i;
        }
        return candidates[lastIdx].i;
    }
}

File: Llama3.Rust/src/sampler.rs
use rand::Rng;
use crate::tensors::softmax;

pub struct Sampler {
    rng: rand::rngs::StdRng,
}

impl Sampler {
    pub fn new(seed: Option<u64>) -> Self {
        let rng = if let Some(s) = seed {
            rand::SeedableRng::seed_from_u64(s)
        } else {
            rand::SeedableRng::from_entropy()
        };
        Sampler { rng }
    }

    pub fn sample(&mut self, logits: &mut [f32], temp: f32, topp: f32) -> u32 {
        if temp == 0.0 {
            return logits.iter().enumerate()
                .max_by(|(_, a), (_, b)| a.partial_cmp(b).unwrap())
                .map(|(i, _)| i as u32)
                .unwrap();
        }

        for v in logits.iter_mut() { *v /= temp; }
        softmax(logits);

        if topp <= 0.0 || topp >= 1.0 {
            let r: f32 = self.rng.gen();
            let mut cdf = 0.0;
            for (i, &p) in logits.iter().enumerate() {
                cdf += p;
                if r < cdf { return i as u32; }
            }
            return (logits.len() - 1) as u32;
        }

        self.sample_topp(logits, topp)
    }

    fn sample_topp(&mut self, logits: &[f32], topp: f32) -> u32 {
        let threshold = (1.0 - topp) / (logits.len() as f32);
        let mut candidates: Vec<(usize, f32)> = logits.iter().enumerate()
            .filter(|&(_, &p)| p > threshold)
            .map(|(i, &p)| (i, p))
            .collect();

        if candidates.is_empty() {
            return logits.iter().enumerate()
                .max_by(|(_, a), (_, b)| a.partial_cmp(b).unwrap())
                .map(|(i, _)| i as u32)
                .unwrap();
        }

        candidates.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap());

        let mut cumulative_prob = 0.0;
        let mut last_idx = candidates.len() - 1;
        for (i, &(_, p)) in candidates.iter().enumerate() {
            cumulative_prob += p;
            if cumulative_prob > topp {
                last_idx = i;
                break;
            }
        }

        let r: f32 = self.rng.gen::<f32>() * cumulative_prob;
        let mut cdf = 0.0;
        for &(i, p) in candidates[..=last_idx].iter() {
            cdf += p;
            if r < cdf { return i as u32; }
        }
        candidates[last_idx].0 as u32
    }
}
